### ISSUE TITLE
How to train and evaluate mono phoneme model?

### ISSUE THREAD

**Issue by Caoang327** [mravanelli/pytorch-kaldi#48]
Hi. Thanks for your work on such an amazing package.
I have trained and tested tri-phone model and it worked pretty well. Right now I hope to train a DNN model using mono phoneme model per frame. I change the cfg file as you mentioned and set the output number of MLP as 48. I finish the train and validating phrase. But when I try to decode it, an error occurred saying that "48 mismatched with 144". 

So I wonder how could I set the training and decode the output of Mono phone DNN? I guess 144 may refer to the 144 states, but I don't understand why I need these states for mono phoneme label per frame. And if we don't need to decode for mono phoneme case, how could I evaluate on test data?

Thanks

**Comment by mravanelli** [comment]
Hi,
are you using TIMIT? Which is the decoding graph you are using?

Mirco

**Comment by Caoang327** [comment]
Hi, I am using TIMIT. The setting is:
lab_folder=/home/ubuntu/kaldi/egs/timit/s5/exp/mono_ali
        lab_opts=ali-to-phones --per-frame=true
        lab_count_file=none
        lab_data_folder=/home/ubuntu/kaldi/egs/timit/s5/data/train/
        lab_graph=/home/ubuntu/kaldi/egs/timit/s5/exp/mono/graph
Mono is generated by kaldi s5 run.sh

**Comment by mravanelli** [comment]
Hi,
this is a weird setup (usually the decoding is done with context-dependent
graphs). Is there a particular reason why you are using it?
However, since you are directly using monophone labels you should use:
lab_opts=ali-to-pdf
you also have to set the number of output to 144. 144 is the number of
state (each of the 48 phonemes are described by 3 states).

Please, let me know if this solves the problem...

Mirco

**Comment by Caoang327** [comment]
Hi. I hope to translate the output of DNN by wireless communication so I hope the output number to be small enough. And I'm curious about the performance between mono phone and CTC based strategy.

I'm curious about your "directly using monophone labels". Does it mean I could also set lab_folder=/home/ubuntu/kaldi/egs/timit/s5/exp/dnn4_pretrain-dbn_dnn_ali and use lab_opts=ali-to-phones --per-frame=true? If it is feasible, how should I set graph? exp/mono/graph or exp/tri3/graph?

If I set the number of output to 144, how should I set the loss_mono=cost_nll(out_dnn3,lab_mono)? I guess lab_mono may be 48 dim. So how could I write this part?

Thank you very much

**Comment by mravanelli** [comment]
To decode with monopone you should use a monophone graph and I think the
right way is what I explained in the previous mail. Let me know if it works.
Which of the available examples are you talking about? In the available
examples the monophones are onyl use to regulariza the DNN with the
so-called monophone reguarization, but the decoding is done with
context-dependent models..

Mirco

**Comment by Caoang327** [comment]
Thank you. I will try it.

**Comment by mravanelli** [comment]
you are welcome!
